Add initial requirements for integration test (#630)
open-telemetry/opentelemetry-specification: specification/library-guidelines.md

5. The SDK implementation should include the following exporters:
    - Jaeger.
    - Zipkin.

>     - OpenCensus.

Reviewer: This seems unrelated, please send as a separate PR.

Author: Sent a separate PR #631.

See the [Concurrency and Thread-Safety](concurrency.md) specification for
guidelines on what concurrency safeties should API implementations provide
and how they should be documented.

### Integration Test

The goal of integration test is to make sure:
* The API package and SDK package are designed properly, so another SDK can be used to work with the API package.

> * The exporters are functional, and the emitted data meets with the expectation.

Reviewer 2: ```suggestion
* The exporters are functional, and the emitted data meets expectations.
```

* The context propagation is implemented correctly, so a distributed trace could be built among multiple services.
* Sampling logic is implemented consistently across the services, and across the SDKs.

A test harness is needed to verify each language library's integration and interoperability with other libraries and backends.

Each language library should use the test harness as part of the CI (continuous integration).

> #### Test Harness

Reviewer 2: In OpenTracing we had a separate test harness framework that was operating purely at the API level. The implementation test had to provide certain introspection capabilities to the framework to validate the behavior (fairly limited since OT was not very prescriptive about the actual behavior). This was especially useful for dynamic languages.

* https://github.com/opentracing/opentracing-javascript/blob/master/src/test/api_compatibility.ts
* https://github.com/opentracing/opentracing-go/blob/master/harness/api_checkers.go

See the [Concurrency and Thread-Safety](concurrency.md) specification for
guidelines on what concurrency safeties should API implementations provide
and how they should be documented.

### Integration Test

> The goal of integration test is to make sure:

Reviewer 3: ```suggestion
The goal of the integration test is to make sure:
```
(or plural?)

> * The API package and SDK package are designed properly, so another SDK can be used to work with the API package.

Reviewer 3: ```suggestion
* The API package and SDK package are designed properly, so that an SDK other than the default SDK can be used to work with the API package.
```

* The exporters are functional, and the emitted data meets expectations.
* The context propagation is implemented correctly, so a distributed trace could be built among multiple services.

> * Sampling logic is implemented consistently across the services, and across the SDKs.

Reviewer 3: Across which services?

> A test harness is needed to verify each language library's integration and interoperability with other libraries and backends.

Reviewer 3: Which other libraries? Do you mean other OpenTelemetry implementations (API + default SDK) in other processes?

> A test harness is needed to verify each language library's integration and interoperability with other libraries and backends.

Reviewer 3: ```suggestion
A test harness is needed to verify each language library's integration and interoperability with other libraries and certain backends listed below.
```

Each language library should use the test harness as part of the CI (continuous integration).

#### Test Harness

The test harness should include:
* A document which explains how to use it to implement integration test.

> * A [Docker compose](https://docs.docker.com/compose/).

Reviewer 3: ```suggestion
* A [Docker compose](https://docs.docker.com/compose/) file.
```

> * A container image which provides integration test service that can be invoked from the library's CI job.

Reviewer 3: ```suggestion
* A container image which provides an integration test service that can be invoked from the library's CI job.
```

> * A simplied version of the OpenTelemetry C++ SDK, which will be used to test the interfacing between API and SDK. For example, using OpenTelemetry Python API with C++ SDK.

Reviewer 3: ```suggestion
* A simplified version of the OpenTelemetry C++ SDK, which will be used to test the interfacing between API and SDK. For example, using the OpenTelemetry Python API with the C++ SDK.
```

* The latest version of OpenTelemetry Collector.
* The latest stable version of Prometheus backend.
* The latest stable version of Jaeger backend.
* The latest stable version of Zipkin backend.

#### CI Requirement

Each language library should implement a test service which interacts with the test harness. The test service should include:

> * A REST endpoint which is used to receive incoming requests from the harness and other services that are involved during the CI.

Reviewer 3: Please clarify what you mean with "other services that are involved during the CI". Currently it sounds like this was about CI services like CircleCI or AppVeyor.

* The ability to react to the requests from the harness.
* The ability to specify the SDK.

>   * Without the SDK

Reviewer 3: ```suggestion
  * Without the SDK (API-only mode)
```
Note that currently the behavior differs widely between language implementations, see https://github.com/open-telemetry/opentelemetry-specification/issues/520.

  * With the language specific SDK
  * With the simplifed C++ SDK

> * The ability to configure the SDK, when the language specific SDK is used.

Reviewer 3: ```suggestion
* The ability to configure the SDK, when the language specific default SDK is used.
```

  * Sampler
  * Exporter
* The ability to call other REST endpoints based on the instruction given by the harness.

> Each language SDK should determine which platform and runtime version should be tested. For example, a language library could decide to only cover one specific combination of platform and runtime during integration test, and rely on the unit test to cover the other platforms and runtime versions.

Reviewer 3: ```suggestion
Each language SDK should determine which platform and runtime version should be tested. For example, a language library could decide to only cover one specific combination of platform and runtime during integration test, and rely on the unit tests to cover the other platforms and runtime versions.
```

Hopefully we have more than one 😁
